### DISCUSSION TITLE
What does `fit` do when calculating realized performance?

### DISCUSSION THREAD

**Discussion by shezadkhan137** [NannyML/nannyml#336] [Q&A] [answered]
Hi, Thanks for all of your work on Nannyml! 

We want to use nannyml to write realized performance data into a SQL database, and then use Grafana to plot these metrics over time. My question is is there any way to use the `PerformanceCalculator` without having to call `fit` first? I'm not exactly sure what fit is doing, and it seems to be quite slow, even on small amounts of data.

**Comment by santiviquez** [comment, marked as answer]
Hi @shezadkhan137!

When using the `PerformanceCalculator` method, the purpose of the `fit` function is to create the baseline that represents the realized performance on the reference set. This is where nannyML calculates the thresholds that can later be useful in determining whether there has been degradation on the analysis set, like in the image below.

<img width="611" alt="Screenshot 2023-11-15 at 11 02 14" src="https://github.com/NannyML/nannyml/assets/10890881/68585050-5fc9-4e9c-a439-13c6f777feb4">

Regarding your question about using `PerformanceCalculator` with out calling `fit` first, actually yes, you can.

You only need to call `fit` when the reference data changes. So, if your reference data hasn't change, you can call fit only once, persist the fitted calculator (we currently support multiple ways to persist calculator objects), and then use it to perform calculations whenever needed. This should reduce the running time.

I'll let you here a [tutorial on how to store and load calculators](https://nannyml.readthedocs.io/en/stable/tutorials/storing_and_loading_calculators.html) 

I hope this helps :) Let me know if you have any questions!

**Reply by nnansters** [reply]
Additional bit of information. The duration of fitting will depend on multiple factors:

- Size of the reference data set: the larger it is, the longer fitting will take. 
- Chunking: the more chunks you end up with, the more computations will be executed during fitting, so the longer it will take.
- The operation itself. This should not take too much time for the realized performance calculation. 

That being said, the intended flow is indeed to fit once and reuse the fitted calculator as long as the reference data doesn't change (too much).
